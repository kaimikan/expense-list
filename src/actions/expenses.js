import uuid from "uuid";
import db from "../firebase/firebase";
import {
  getDatabase,
  ref,
  get,
  child,
  set,
  update,
  remove,
  onValue,
  off,
  push,
  onChildAdded,
  onChildChanged,
  onChildRemoved
} from "firebase/database";

// PRE ASYNCHRONOUS
/* 
-component calls action generator
-action generator returns object
-component distaptches object
-redux store changes
*/

//POST ASYNCHRONOUS
/* 
-components calls action generator
-action generator returns function
-component dispatches function (?)
-function runs (has the ability to dispatch other actions and do whatever it wants)
*/

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  // this syntax works because of redux thunk
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return push(ref(db, "expenses"), expense).then((ref) => {
      dispatch(
        addExpense({
          id: ref.key,
          ...expense
        })
      );
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ expenseIDToRemove } = {}) => ({
  type: "REMOVE_EXPENSE",
  expenseIDToRemove
});

export const startRemoveExpense = ({ expenseIDToRemove } = {}) => {
  return (dispatch) => {
    return remove(ref(db, `expenses/${expenseIDToRemove}`)).then(() => {
      dispatch(removeExpense({ expenseIDToRemove }));
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  const updatesHolder = {};
  updatesHolder[`expenses/${id}`] = { ...updates };
  return (dispatch) => {
    return update(ref(db), updatesHolder).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses
});

//export const startSetExpenses;
export const startSetExpenses = () => {
  // this syntax works because of redux thunk
  return (dispatch) => {
    return get(child(ref(db), `expenses`)).then((dataSnapshot) => {
      const expensesData = [];
      dataSnapshot.forEach((childSnapshot) => {
        const id = childSnapshot.key;
        expensesData.push({
          id,
          ...childSnapshot.val()
        });
      });

      return dispatch(setExpenses(expensesData));
    });
  };
};
