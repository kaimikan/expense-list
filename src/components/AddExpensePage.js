import React from "react";

const AddExpensePage = () => <div>This is from the add component</div>;

export default AddExpensePage;
