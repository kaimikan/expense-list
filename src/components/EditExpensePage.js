import React from "react";

const EditExpensePage = (props) => {
  console.log("props");
  return <div>This is from the edit component {props.match.id}</div>;
};
export default EditExpensePage;
