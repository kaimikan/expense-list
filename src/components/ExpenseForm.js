import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

//const date = new Date();
const now = moment();
console.log(now.format("Do MMM, YYYY"));

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    // we use the constructor to be able to handle default info for both edit and add expense functionalities
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ""
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note
    }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    // you can check the match() regex on regex101.com
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) this.setState(() => ({ createdAt }));
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  // if i used the name onSubmit it wasn't going through (it seems to be a reserved name)
  // nevermind it works now but not sure why
  onSubmit = (e) => {
    // prevent page refresh
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide description and amount"
      }));
    } else {
      this.setState(() => ({
        error: ""
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && (
          <p className="form__error">
            <b>Error: </b> {this.state.error}
          </p>
        )}
        <input
          type="text"
          className="text-input"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          className="text-input"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={(/* day */) => false}
          firstDayOfWeek={1}
          displayFormat="D/MM/YYYY"
        />
        {/* created at will be done with a third-party component */}
        <textarea
          className="textarea"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
          <button className="button">Save Expense</button>
        </div>
      </form>
    );
  }
}
