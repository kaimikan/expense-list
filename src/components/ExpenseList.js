import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {props.expenses.map((expense) => {
      return (
        <div>
          <b>Expense Item</b>
          <ExpenseListItem key={expense.id} {...expense} />
        </div>
      );
    })}
  </div>
);

/* Long Version
const ConnectedExpenseList = connect((state) => {
  return {
    expenses: state.expenses
  };
})(ExpenseList);

export default ConnectedExpenseList;
*/

// Short Version
const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
