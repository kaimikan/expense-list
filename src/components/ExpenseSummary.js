import React from "react";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";
import sumExpensesAmounts from "../selectors/expenses-total";
import numeral from "numeral";

export const ExpenseSummary = (props) => (
  <div>
    {/* would be smarter if the ternary operator was concluded in variables instead of having
     multiple paragraphs being reused along with the function calls
     - will keep it to commemorate an nonoptimal solution
     e.g. const expenseText = props.numberOfExpenses === 1 ? 'expense' : 'expenses';
     */}
    {props.numberOfExpenses === 0 ? (
      <p></p>
    ) : props.numberOfExpenses === 1 ? (
      <p>
        Viewing 1 expense totalling €
        {numeral(props.expensesTotalAmount / 100).format("0,0[.]00")}
      </p>
    ) : (
      <p>
        Viewing {props.numberOfExpenses} expenses totalling €
        {numeral(props.expensesTotalAmount / 100).format("0,0[.]00")}
      </p>
    )}
  </div>
);

// Short Version
const mapStateToProps = (state) => {
  return {
    numberOfExpenses: selectExpenses(state.expenses, state.filters).length,
    expensesTotalAmount: sumExpensesAmounts(state.expenses)
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
