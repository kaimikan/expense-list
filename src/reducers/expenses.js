// Expenses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      // WAY #1
      // // concat does not change the state variable, makes a copy instead and then changes it
      // return state.concat(action.expense);

      // WAY #2 ARRAY SPREAD OPERATOR ...
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(
        (expenseID) => action.expenseIDToRemove !== expenseID.id
      );
    case "EDIT_EXPENSE":
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            // OBEJCT SPREAD OPERATOR ...
            // we take all of the indivial Expense's values and overwrite
            // only the ones which action.updates has
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};
