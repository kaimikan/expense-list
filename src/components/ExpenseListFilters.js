import React from "react";
import { connect } from "react-redux";
import { setTextFilter } from "../actions/filters";

const ExpenseListFilters = (props) => (
  <div>
    <input
      type="text"
      value={props.filters.text}
      onChange={(e) => {
        // same as store.dispatch(...)
        props.dispatch(setTextFilter(e.target.value));
        console.log(e.target.value);
      }}
    />
  </div>
);

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

// connect helps redux use the state from react as props basically
export default connect(mapStateToProps)(ExpenseListFilters);
