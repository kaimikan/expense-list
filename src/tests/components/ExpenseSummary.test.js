import React from "react";
import { shallow } from "enzyme";
import { ExpenseSummary } from "../../components/ExpenseSummary";
import expenses from "../fixtures/expenses";

test("should render ExpenseSummary with multiple expenses", () => {
  const wrapper = shallow(<ExpenseSummary numberOfExpenses={2} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary with 1 expense", () => {
  const wrapper = shallow(<ExpenseSummary numberOfExpenses={1} />);
  expect(wrapper).toMatchSnapshot();
});
