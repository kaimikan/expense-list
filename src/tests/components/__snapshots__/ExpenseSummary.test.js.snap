// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with 1 expense 1`] = `
<div>
  <p>
    Viewing 1 expense totalling €
    0
  </p>
</div>
`;

exports[`should render ExpenseSummary with multiple expenses 1`] = `
<div>
  <p>
    Viewing 
    2
     expenses totalling €
    0
  </p>
</div>
`;
